test/controllers/close_contacts_controller_test.rb

```diff
@@ -0,0 +1,203 @@
+
+    put :update, params: { id: 'test', patient_id: 'test' }
+    assert_response(:bad_request)
+    assert_equal("Close Contact cannot be modified for unknown monitoree with ID: #{'test'.to_i}", JSON.parse(response.body)['error'])
+
+    put :destroy, params: { id: 'test', patient_id: 'test' }
+    assert_response(:bad_request)
+    assert_equal("Close Contact cannot be modified for unknown monitoree with ID: #{'test'.to_i}", JSON.parse(response.body)['error'])
+
+    sign_out user
+  end
+
+  test 'before action: check patient (current user can view patient)' do
+    user = create(:public_health_enroller_user)
+    user_2 = create(:public_health_enroller_user)
+    patient = create(:patient, creator: user_2)
+    sign_in user
+
+    post :create, params: { patient_id: patient.id }
+    assert_response(:forbidden)
+    assert_equal("User does not have access to Patient with ID: #{patient.id}", JSON.parse(response.body)['error'])
+
+    put :update, params: { id: 'test', patient_id: patient.id }
+    assert_response(:forbidden)
+    assert_equal("User does not have access to Patient with ID: #{patient.id}", JSON.parse(response.body)['error'])
+
+    put :destroy, params: { id: 'test', patient_id: patient.id }
+    assert_response(:forbidden)
+    assert_equal("User does not have access to Patient with ID: #{patient.id}", JSON.parse(response.body)['error'])
+
+    sign_out user
+  end
+
+  test 'before action: check close_contact' do
+    user = create(:public_health_enroller_user)
+    patient = create(:patient, creator: user)
+    sign_in user
+
+    put :update, params: { id: 'test', patient_id: patient.id }
+    assert_response(:bad_request)
+
+    put :destroy, params: { id: 'test', patient_id: patient.id }
+    assert_response(:bad_request)
+
+    sign_out user
+  end
+
+  # --- CREATE --- #
+
+  test 'create: creates new close contact and creates related history item' do
+    user = create(:public_health_enroller_user)
+    patient = create(:patient, creator: user)
+
+    sign_in user
+    first_name = 'test'
+    last_name = 'test'
+    email = 'email@test.com'
+    post :create, params: {
+      first_name: first_name,
+      last_name: last_name,
+      email: email,
+      patient_id: patient.id
+    }
+
+    assert_response(:success)
+    assert_equal(1, patient.close_contacts.count)
+    assert_equal(1, patient.histories.count)
+
+    close_contact = patient.close_contacts.first
+    assert_equal(first_name, close_contact[:first_name])
+    assert_equal(last_name, close_contact[:last_name])
+    assert_equal(email, close_contact[:email])
+
+    history = patient.histories.first
+    assert_equal(History::HISTORY_TYPES[:close_contact], history[:history_type])
+    assert_equal(user.email, history[:created_by])
+    assert_equal("User added a new close contact (ID: #{close_contact.id}).", history[:comment])
+
+    sign_out user
+  end
+
```

Comment:
I'd like to see tests (one for each create, update, destroy) for if those actions fail.  You'll probably have to restructure the controller slightly to do so.  See the labs controller and test file for what I am talking about

Reply:
Also, once you restructure the controller, wrap the save/update destroy in a transaction.  Again, see the labs controller for examples